addons/nuxeo-retention/nuxeo-retention-core/src/main/java/org/nuxeo/retention/service/RetentionManagerImpl.java

```diff
@@ -0,0 +1,353 @@
+        ctx.setProperty(DocumentEventContext.CATEGORY_PROPERTY_KEY, DocumentEventCategories.EVENT_DOCUMENT_CATEGORY);
+        ctx.setProperty(DocumentEventContext.COMMENT_PROPERTY_KEY, rule.getDocument().getPathAsString());
+        Event event = ctx.newEvent(RetentionConstants.RULE_ATTACHED_EVENT);
+        Framework.getService(EventService.class).fireEvent(event);
+    }
+
+    protected void checkCanAttachRule(DocumentModel document, RetentionRule rule, CoreSession session) {
+        NuxeoPrincipal principal = session.getPrincipal();
+        if (!principal.isAdministrator() && !principal.isMemberOf(RetentionConstants.RECORD_MANAGER_GROUP_NAME)) {
+            if (!session.hasPermission(document.getRef(), SecurityConstants.MAKE_RECORD)
+                    || !session.hasPermission(document.getRef(), SecurityConstants.SET_RETENTION))
+                throw new NuxeoException("User is not authorized to attach retention rule", SC_FORBIDDEN);
+        }
+        if (!rule.isEnabled()) {
+            throw new NuxeoException(String.format("Rule is disabled"));
+        }
+        if (!rule.isDocTypeAccepted(document.getType())) {
+            throw new NuxeoException("Rule does not accept this document type");
+        }
+        if (document.hasFacet(RetentionConstants.RECORD_FACET)) {
+            throw new NuxeoException("Document is already a record");
+        }
+    }
+
+    @Override
+    public boolean canAttachRule(DocumentModel document, RetentionRule rule, CoreSession session) {
+        try {
+            checkCanAttachRule(document, rule, session);
+            return true;
+        } catch (NuxeoException e) {
+            log.info("Cannot attach rule {} on document {}", () -> rule.getDocument().getPathAsString(),
+                    document::getPathAsString);
+            return false;
+        }
+    }
+
+    public void executeRuleBeginActions(Record record, CoreSession session) {
+        RetentionRule rule = record.getRule(session);
+        if (rule != null) {
+            executeRuleActions(record.getDocument(), rule.getBeginActions(), session);
+        }
+    }
+
+    public void executeRuleEndActions(Record record, CoreSession session) {
+        RetentionRule rule = record.getRule(session);
+        if (rule != null) {
+            executeRuleActions(record.getDocument(), rule.getEndActions(), session);
+        }
+    }
+
+    protected void executeRuleActions(DocumentModel doc, List<String> actionIds, CoreSession session) {
+        if (actionIds != null) {
+            AutomationService automationService = Framework.getService(AutomationService.class);
+            for (String operationId : actionIds) {
+                log.debug("Executing {} action on {}", () -> operationId, doc::getPathAsString);
+                // Do not lock document if already locked, and unlock if already unlocked (triggers an error)
+                // Also, if it's time to delete, unlock it first, etc.
```

Comment:
i think you mean `/ Do not lock document if already locked, and unlock if already unlocked (triggers an err...` -> 
`Do not lock document if already locked, and lock if already unlocked...` or i am wrong

Reply:
Replaced with clearer `Do not lock document if already locked, nor unlock if already unlocked (would trigger an error)`